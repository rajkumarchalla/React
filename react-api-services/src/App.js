import React, { Component } from 'react';
import './App.css';
import { getuserData } from './APIUtils';
import constant from './Constant'

class App extends Component {
  /**
   * 
   * @param {*} props 
   */
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
    this.state = {
      isLoading: false,
      contacts: [], //contacts array
      show: false, // sate flag,
      data: []
    };
    this.updateItem = this.updateItem.bind(this);
  }
  //Load data from API action 
  handleClick() {
    this.setState({ show: true }, () => {
    });
    this.setState({ isLoading: true }, () => {
      simulateNetworkRequest().then(() => {
        this.setState({ isLoading: false });
      });
    });
  }
  /** #MARK:- Component LifeCycle Method */
  componentDidMount() {
    /**
     * {*} params - parameters 
     */
    const params = {
      name: "Kurtis Weissnat",
      username: "Elwyn.Skiles"
    }
    /** get user data */
    getuserData(constant.req1UrlParams, params)
      .then(response => {
       // console.log(response)
        this.updateItem(response)
      })
      .catch(errorMessage => {
        alert(errorMessage)
        console.log(errorMessage)
      });

  }
  /** updateItem as a prop to the child Component.
   * 
   * @param {*} res - Resposne 
   */
  updateItem(res) {
    //console.log('Selected Value:: ', res);
    // create an array of contacts only with relevant data
    const newContacts = res.data.map(c => {
      return {
        id: c.id,
        name: c.name,
        email: c.email
      };
    });
    // create a new "State" object without mutating 
    // the original State object. 
    const newState = Object.assign({}, this.state, { contacts: newContacts });
    // store the new state object in the component's state
    this.setState(newState);
    this.setState({ isLoading: false });
    this.setState({ show: false });
  }
  /**
   * @React - Render 
   *  <h1> React API Layer View Sample </h1> 
   *         <div>
              <NumberDisplay number={2}> </NumberDisplay>
            </div>
   */
  render() {
    const { isLoading } = this.state;
    return (
      <div className="container">
        <div className="row">
          <div className="col-sm">
            <div >
              <h1> Front-End-SDK TDD </h1>
            </div>
            <br />

            <br />
          </div>
        </div>
      </div>
    );
  }
}
/**
 * @function - simulateNetworkRequest
 */
//Sample delay using promise 
function simulateNetworkRequest() {
  return new Promise(resolve => setTimeout(resolve, 2000));
}
export default App;
