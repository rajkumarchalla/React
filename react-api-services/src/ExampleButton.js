import React from "react";
import PropTypes from "prop-types";
import styles from "./styles.css";
import Button from 'react-bootstrap/Button'

const propTypes = {
    children: PropTypes.node,
    multi: PropTypes.bool,
    handle: PropTypes.func,
    onClick: PropTypes.func,
    tag: PropTypes.tag,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    animated: PropTypes.bool,
    color: PropTypes.string,
    className: PropTypes.string,
    barClassName: PropTypes.string,
    cssModule: PropTypes.object,
    btTitle: PropTypes.string,
    type: PropTypes.string,
    close: PropTypes.bool,
    active: PropTypes.bool
};
const defaultProps = {
    tag: "div",
    value: 0,
    text: "Sample bttttn",
    title : "ccccc"
};
const ExampleButton = props => {
    const {
        className,
        barClassName,
        cssModule,
        value,
        max,
        animated,
        striped,
        color,
        lableColor,
        close,
        onClick,
        active,
        circular,
        tag: Tag,
        btTitle,
        type,
        ...attributes
    } = props;


    const ExampleButtonP = (

        <div className="container">
            <div className="row">
                <div className="col-sm">
                    <span className="default" style={{ color: lableColor }}>
                    <h1> Button Test </h1>
                    <button className={styles.buttonStyle}  color={props.lableColor} type={type} onClick={onClick}  lableColor={styles.buttonStyle} > {btTitle} </button> <br/> 
                    <Button className={styles.buttonStyle} variant="outline-primary" size="lg" block onClick={onClick} > {btTitle} </Button>
                    </span>
                </div>
            </div>
        </div>
    );
    if (ExampleButton) {
        return ExampleButtonP;
    }
    return (
        <Tag {...attributes} className={ExampleButton} children={ExampleButton} />
    );
};

ExampleButton.propTypes = propTypes;
ExampleButton.defaultProps = defaultProps;

export default ExampleButton;