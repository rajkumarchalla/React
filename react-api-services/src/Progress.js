import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import {toNumber} from "lodash";
import { mapToCssModules, tagPropType } from "./utils";
import styles from "./styles.css";

const propTypes = {
  children: PropTypes.node,
  bar: PropTypes.bool,
  multi: PropTypes.bool,
  handle: PropTypes.func,
  tag: tagPropType,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  max: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  animated: PropTypes.bool,
  striped: PropTypes.bool,
  color: PropTypes.string,
  className: PropTypes.string,
  barClassName: PropTypes.string,
  title: PropTypes.string,
  cssModule: PropTypes.object,
  onClick: PropTypes.func
};

const defaultProps = {
  tag: "div",
  value: 0,
  max: 100,
  animated: false
};

const Progress = props => {
  const {
    children,
    className,
    barClassName,
    cssModule,
    value,
    max,
    animated,
    striped,
    color,
    lableColor,
    bar,
    circular,
    tag: Tag,
    title,
    onClick,
    ...attributes
  } = props;

  const percent = (toNumber(value) / toNumber(max)) * 100;

  const progressClasses = mapToCssModules(
    classNames(
      className
      // 'progress'
    ),
    cssModule
  );

  const progressBarClasses = mapToCssModules(
    classNames(
      "progress-bar",
      bar ? className || barClassName : barClassName,
      animated ? "progress-bar-animated" : null,
      color ? `bg-${color}` : null,
      striped || animated ? "progress-bar-striped" : null
    ),
    cssModule
  );

  const circularProgressBarClasses = classNames(
    classNames(
      animated ? "progress-bar-animated" : null,
      color ? `bg-${color}` : null,
      striped || animated ? "progress-bar-striped" : null
    ),
    cssModule
  )


  const circularProgressBarLabelClasses = classNames(
    
  )

  const ProgressBar = circular ? (
    <div className={styles.progress} onClick={onClick} >
      <span className={styles.progressleft}>
        <span className={styles.progressbar}></span>
      </span>
      <span className={styles.progressright}>
        <span className={styles.progressbar}></span>
      </span>
      <div className={styles.progressvalue} style={{ color: lableColor} }>90%</div>
</div>
  ) : (
    <div className="container" onClick={onClick}>
      <div className="row">
        <div className="col-sm">
          <span className={styles.ovlabel} style={{ color: lableColor} } data-testid="progress-displa">
            {title}
          </span>
        </div>
        <div className="col-sm" >
          <div className="progress">
          <span data-testid="progress-display">
            <div
              className={progressBarClasses}
              style={{ width: `${percent}%` }}
              role="progressbar"
              aria-valuenow={value}
              aria-valuemin="0"
              aria-valuemax={max}
              children={children}
            />
            </span>
          </div>
        </div>
      </div>
    </div>
  );

  if (bar) {
    return ProgressBar;
  }

  return (
    <Tag {...attributes} className={progressClasses} children={ProgressBar} />
  );
};

Progress.propTypes = propTypes;
Progress.defaultProps = defaultProps;

export default Progress;
