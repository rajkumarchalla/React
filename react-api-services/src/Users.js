import React, { Component } from "react";
import { getuserData } from './APIUtils';
import constant from './Constant'

export default class Users extends Component {
  constructor(props) {
    super(props);
    this.state = { data: [] };
    this.handleGet = this.handleGet.bind(this);
  }
  handleGet(response) {
    this.setState(this.stateUpdater(response));
  }
  stateUpdater(response) {
    return { data: response.data };
  }
  async componentDidMount() {
    const params = {
      name: "Kurtis Weissnat",
      username: "Elwyn.Skiles"
    }
    /** get user data */
    await getuserData(constant.req1UrlParams, params)
      .then(response => {
       // console.log(response)
        this.handleGet(response);
      })
      .catch(errorMessage => {
        alert(errorMessage)
       // console.log(errorMessage)
      });
  }
  render() {
    return (
      <ul>
        {this.state.data.map(user => (
          <li key={user.name}>{user.name}</li>
        ))}
      </ul>
    );
  }
}