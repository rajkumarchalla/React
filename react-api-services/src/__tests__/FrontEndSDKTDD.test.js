
import React from 'react'
import ReactDOM from 'react-dom'
import { shallow, mount } from 'enzyme';
import Progress from '../Progress'
import App from '../App';
import ExampleButton from '../ExampleButton';
import { create } from 'react-test-renderer';
import 'jest-styled-components';
import styled, { ThemeProvider } from 'styled-components'
import { render } from 'react-testing-library'
import renderer from 'react-test-renderer'
import Users from '../Users';

/* FrontEndSDK TDD
*
* Jest API :- https://jestjs.io/docs/en/configuration
              The Jest API focusses more on the ability to define tests, make assertions, and create mocks.
* describe: defines a test suite.
* it: defines a test.
* beforeEach: defines an entry hook before running each test.
* expect: makes an assertion.
* jest.fn(): creates a mock function.
* toEqual: checks if two objects have the same value.
* toBe: checks if two objects have the same value and type.
* toBeDefined: checks if the object is defined.
* toContain: checks if an item is present in a list.
* toBeCalled: checks if a mock function is called.
*
*----------------------
* Enzyme :- It provides a mechanism to mount and traverse React.js component trees.
          It will help us get access to its own properties and state as well as its children props in order to run our assertions.
          Ref: https://airbnb.io/enzyme/

* Enzyme API focusses on rendering the react component and retrieving specific nodes from the rendered tree. There are three ways to render a component.
* shallow: renders only the component under test. Dependent components are not rendered.
* mount: mounts the full component in JSDOM.
* render: renders the component as static HTML.
*
*----------There are several methods to retrieve nodes from the rendered component.
*
* find: accepts a selector and retrieves nodes that match the selector.
* findWhere: retrieve nodes selected by the predicate.
* some: returns true if there is at-least one node matching the selector.
* someWhere: returns true if there is at-least one node selected by the predicate.
* first: returns the first node of a set.
* at: returns the nth node of a set.
* html: gets the HTML of the node.
* text: gets the text representation of the node.
*
*-----------There are more methods to interact with the component and retrieve the component state.
*
* simulate: simulates an event.
* setProps: sets the props.
* setState: sets the state.
* setContext: sets the context.
* prop(key): retrieves prop value corresponding to the provided key.
* state(key): retrieves state corresponding to the provided key.
* context(key): retrieves context value corresponding to the provided key.
*
*/

describe('TestCase#1 : Should check App Title ', function () {
    it('App\'s title should be React Test Case Sample', function () {
        let app = shallow(<App />);
        expect(app.find('h1').text()).toEqual(' Front-End-SDK TDD ');
    });
});

describe("TestCase#2 : Progress Component Rendering tests", () => {
    it('renders without crashing', () => {
        const div = document.createElement('ProgressVL');
        ReactDOM.render(<Progress />, div);
        ReactDOM.unmountComponentAtNode(div);
    });
});

describe("TestCase#3 :Progress Component Rendering tests", () => {
    //start with most simple test for component render without an error:
    it("TestCase#3 -1 : Should render correctly", () => {
        const component = shallow(<Progress />);
    });

    it("TestCase#3 -2 : Snapshot Test should render initial layout", () => {
        const component = shallow(<Progress />);
        expect(component.getElements()).toMatchSnapshot();
    });
    /** Shallow Rendering */
    it("TestCase#3 -3 : Should render with classname", () => {
        const wrapper = shallow((
            <Progress>
                <div className="progress" />
            </Progress>
        ));
        expect(wrapper.contains(<div className="progress" />)).toEqual(true);
    });

    it('TestCase#3 -4 : Should render with a <div />', () => {
        const wrapper = shallow(<Progress />);
        console.log(wrapper.find('div').length);
        expect(wrapper.find('div').length).toEqual(7);
    });

    it(' TestCase#3 -5 : Should Allows to set props', () => {
        const props = {
            color: ''
        },
            wrapper = mount(<Progress {...props} color='' />);
        wrapper.setProps({ color: 'pcolor' });
        console.log(wrapper.prop('color'));
        expect(wrapper.props().color).toEqual('pcolor');
    });

    it('TestCase#3 -6 :Check the title of progress', () => {
        const props = {
            title: ''
        },
            ProgressComponent = mount(<Progress {...props} title="GV" />);
        console.log(ProgressComponent.prop('title'));
        expect(ProgressComponent.prop('title')).toBe('GV');
    });

    it('TestCase#3 -7 :Check render value with empty', () => {
        const props = {
            value: null
        },
            ProgressComponent = mount(<Progress {...props} />);
        expect((ProgressComponent).prop('value')).toEqual(null);
    });

    it('TestCase#3 -8 :Check the value of progress', () => {
        const props = {
            value: ''
        },
            ProgressComponent = mount(<Progress {...props} value={20} />);
        console.log(ProgressComponent.prop('value'));
        expect(ProgressComponent.prop('value')).toEqual(20);
    });
    it('TestCase#3 -9 : Edge testcase - Check the max value of progress', () => {
        const props = {
            max: '',
            value: ''
        },
            ProgressComponent = mount(<Progress {...props} max={100} value={40} />);
        console.log(ProgressComponent.prop('max'));
        expect(ProgressComponent.prop('value')).toBeLessThanOrEqual(ProgressComponent.prop('max'));
    });

});
describe('TestCase#4 : Test Button click event', () => {

    it('TestCase#4 -1 OnClick ()', () => {
        // here rather than just testing the output we can create a mock function with jest.fn().
        //use Mock jest.fn(), the behavior of a function that is called indirectly by some other code.
        const mockCallBack = jest.fn();
        const button = shallow((<ExampleButton onClick={mockCallBack}>Ok!</ExampleButton>));
        button.find('button').simulate('click');
        expect(mockCallBack.mock.calls.length).toEqual(1);
    });

    function CallBack() {
        this.calls = 0;
        this.name = "InfotraxVL";
    }
    CallBack.prototype.fn = function () {
        return () => this.calls++;
    }
    it('TestCase#4 -2 : Test Button onclick function callback', () => {
        const callBack = new CallBack();
        const mockCallBack = callBack.fn();
        const button = shallow((<ExampleButton onClick={mockCallBack}>Ok!</ExampleButton>));
        button.find('button').simulate('click');
        expect(callBack.calls).toEqual(1);
        expect(callBack.name).toEqual("InfotraxVL");
    });

    function drinkAll(callback, flavour) {
        if (flavour !== 'orange') {
            callback(flavour);
        }
    }
    it('TestCase#4 -3 : Should call start userFlavour on button click', () => {
        const userFlavour = jest.fn(() => { console.log(true) });
        const wrapper = shallow(<ExampleButton onClick={userFlavour} />);
        wrapper.find('button').at(0).simulate('click'); // at- current node of index
        drinkAll(userFlavour, 'lemon');
        expect(userFlavour).toHaveBeenCalled(); //to ensure that a mock function got called.
        console.log(userFlavour());
        drinkAll(userFlavour, 'orange');
        //  expect(userFlavour).not.toHaveBeenCalled();
    });

});

describe('TestCase#5 : Get Data from API', () => {
    it("shows a list of users from API", async () => {
        const component = create(<Users />);
        const instance = component.getInstance();
        await instance.componentDidMount();
        // console.log(instance.state);
    });
});
/** @{Defination} The toHaveStyleRule matcher is useful to test if a given rule is applied to a component.
 * @{args} The first argument is the expected property, the second is the expected value which can be a String, RegExp, Jest asymmetric matcher or undefined.
 * When used with a negated ".not" modifier the second argument is optional and can be omitted.
 * Ref : https://github.com/styled-components/jest-styled-components#tohavestylerule
 * */
describe('TestCase#6 : toHaveStyleRule ', () => {
    const ExampleButton = styled.button`
width: 50%;
height: 30%;
border: 2px solid rgb(20, 233, 162);
color: rgb(18,105,206);
font-size: 24px;
position: relative;
border: 0.05em solid ${props => props.transparent ? 'transparent' : 'black'};
cursor: ${props => !props.disabled && 'pointer'};
opacity: ${props => props.disabled && '.65'};
`
    it("Jest Styled Components works with shallow rendering", () => {
        const wrapper = shallow(<ExampleButton />);
        expect(wrapper.getElements()).toMatchSnapshot();
    });
    it('it applies default styles', () => {

        const tree = renderer.create(<ExampleButton />).toJSON()
        expect(tree).toMatchSnapshot();
        expect(tree).toHaveStyleRule('color', 'rgb(18,105,206)');
        expect(tree).toHaveStyleRule('border', '0.05em solid black');
        expect(tree).toHaveStyleRule('font-size', '24px');
        expect(tree).toHaveStyleRule('cursor', 'pointer');
        expect(tree).not.toHaveStyleRule('opacity'); // equivalent of the following two
        expect(tree).not.toHaveStyleRule('opacity', expect.any(String));
        expect(tree).toHaveStyleRule('opacity', undefined);
    });

    it('it applies styles according to passed props', () => {
        const tree = renderer.create(<ExampleButton disabled transparent />).toJSON();
        expect(tree).toHaveStyleRule('border', expect.stringContaining('transparent'));
        expect(tree).toHaveStyleRule('cursor', undefined);
        expect(tree).toHaveStyleRule('opacity', '.65');
    });
    // it('getDOMNode css ', () => {
    // const wrapper = shallow(<ExampleButton />);
    // const elem = wrapper.find(Element);
    // expect(getComputedStyle(elem.getDOMNode()).getPropertyValue('opacity')).toBe('0.4');
    // })
});
