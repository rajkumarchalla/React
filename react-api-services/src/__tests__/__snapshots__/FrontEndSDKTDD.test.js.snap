// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestCase#3 :Progress Component Rendering tests TestCase#3 -2 : Snapshot Test should render initial layout 1`] = `
Array [
  <div
    className=""
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col-sm"
        >
          <span
            data-testid="progress-displa"
            style={
              Object {
                "color": undefined,
              }
            }
          />
        </div>
        <div
          className="col-sm"
        >
          <div
            className="progress"
          >
            <span
              data-testid="progress-display"
            >
              <div
                aria-valuemax={100}
                aria-valuemin="0"
                aria-valuenow={0}
                className="progress-bar"
                role="progressbar"
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`TestCase#4 : Test Button click event it works 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-sm"
    >
      <span
        className="default"
        style={
          Object {
            "color": undefined,
          }
        }
      >
        <h1>
           Button Test 
        </h1>
        <button>
           
           
        </button>
         
        <br />
        <button
          className="btn btn-outline-primary btn-block btn-lg"
          disabled={false}
          type="button"
        >
           
           
        </button>
      </span>
    </div>
  </div>
</div>
`;
